design/one-pager-error-and-event-reporting.md

```diff
@@ -0,0 +1,73 @@
+# Error and event reporting
+
+* Owner: Daniel Suskin (@suskin)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Introduction
+
+*Why does this document exist?*
+
+The question of debuggability and troubleshooting has started to come up
+recently, as we have tried out some scenarios that include Stacks and
+different providers. The intent of this document is to propose a mindset
+and framework for reporting errors and other information.
+
+## Goals
+
+In short, a non-admin user and an admin user should both be able to
+debug any issues only by inspecting logs and events. There should be no
+need to rebuild the Crossplane binary or to reach out to a Crossplane
+developer.
+
+A user should be able to:
+
+* Debug an issue without rebuilding the Crossplane binary
+* Understand an issue without contacting a cluster admin
+* Ask a cluster admin to check the logs for more details about the
+  reason the issue happened, if the details are not part of the error
+  message
+
+A cluster admin should be able to:
+
+* Debug an issue without rebuilding the Crossplane binary
+* Debug an issue only by looking at the logs
+* Debug an issue without needing to contact a Crossplane developer
+
+
+## Error reporting in the logs
+
+Error reporting in the logs is mostly intended for consumption by
+cluster admins. A cluster admin should be able to debug any issue by
+inspecting the logs, without needing to add more logs themselves or
+contact a Crossplane developer. This means that logs should contain:
+
+* Error messages
+* Any context leading up to an error, so that the errors can be debugged
+  (the context can be logged at a more verbose level)
+* Ideally, a unique identifier per session so that log lines can be
+  tracked as a sequence when there are multiple workers logging to the
+  same file
+
+## Error reporting as events
+
+Error reporting as Kubernetes events is primarily aimed toward end-users
+of Crossplane who are not cluster admins.
+[Events](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#event-v1-core)
+are available as top-level Kubernetes objects, and show up on objects
+they are related to.
+
+Events should be recorded in the following cases:
```

Comment:
@negz and I also discussed outside of this pull request how it would be useful to also propose changes to the examples in the services developer guide, to help clarify what this would look like in practice. I plan to add changes like that to this pull request.

Reply:
I ended up not having time to add that as part of this pull request; the current plan is covered in [this comment](https://github.com/crossplaneio/crossplane/pull/858#issuecomment-543387834) and in issue #944